File: leetcode/0976_largest_perimeter_triangle.cpp
// Copyright (c) 2019 Jacopo Notarstefano

#include <algorithm>
#include <cassert>
#include <functional>
#include <vector>

using namespace std;

class Solution {
 public:
  int largestPerimeter(vector<int>& A) {
    sort(A.begin(), A.end(), greater<int>());

    for (int i = 0; i < A.size() - 2; i++) {
      if (A[i + 2] + A[i + 1] > A[i]) {
        return A[i + 2] + A[i + 1] + A[i];
      }
    }

    return 0;
  }
};

int main() {
  auto solution = Solution();

  vector<int> A = {2, 1, 2};

  assert(5 == solution.largestPerimeter(A));
}

File: leetcode/0976_largest_perimeter_triangle.py
# -*- coding: utf-8 -*-


class Solution:
    def largestPerimeter(self, A):
        A.sort(reverse=True)
        for i in range(len(A) - 2):
            if A[i + 2] + A[i + 1] > A[i]:
                return A[i + 2] + A[i + 1] + A[i]
        return 0


if __name__ == '__main__':
    solution = Solution()

    assert 5 == solution.largestPerimeter([2, 1, 2])
    assert 0 == solution.largestPerimeter([1, 2, 1])
    assert 10 == solution.largestPerimeter([3, 2, 3, 4])
    assert 8 == solution.largestPerimeter([3, 6, 2, 3])
